src/components/structures/TimelinePanel.js

```diff
@@ -1119,6 +1130,73 @@ const TimelinePanel = createReactClass({
         };
     },
 
+    /**
+     * Check for undecryptable messages that were sent while the user was not in
+     * the room.
+     *
+     * @param {Array<MatrixEvent>} events The timeline events to check
+     * @param {Room} room The room that the events are in
+     *
+     * @return {Number} The index within `events` of the event after the most recent
+     * undecryptable event that was sent while the user was not in the room.  If no
+     * such events were found, then it returns 0.
+     */
+    _checkForPreJoinUISI: function(events) {
+        const room = this.props.timelineSet.room;
+
+        if (events.length === 0 || !room ||
+            !MatrixClientPeg.get().isRoomEncrypted(room.roomId)) {
+            return 0;
+        }
+
+        const userId = MatrixClientPeg.get().credentials.userId;
+
+        let i;
+        let userMembership = "leave";
+        // find the last event that is in a timeline (events may not be in a
+        // timeline if they have not been sent yet) and get the user's membership
+        // at that point.
+        for (i = events.length - 1; i >= 0; i--) {
+            const timeline = room.getTimelineForEvent(events[i].getId());
```

Comment:
Hmm, I was hoping that moving up to `TimelinePanel` would also allow removing the outer loop and timeline check by only looking at `liveEvents` in state (instead of plain `events`). Is it possible to make that simplification, or does it break down somewhere?

Reply:
You're right, we don't need to use events.  I've changed it to use liveEvents instead of events.  (liveEvents vs events confused me.) I'm not sure if the outer loop is needed, but if it isn't, then it will exit after the first iteration of the loop.